File: backend/src/main/java/com/travelcompanion/service/CompanionService.java
package com.travelcompanion.service;

import com.travelcompanion.dto.CompanionDTO;
import com.travelcompanion.entity.TravelCompanion;
import com.travelcompanion.entity.User;
import com.travelcompanion.repository.TravelCompanionRepository;
import com.travelcompanion.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CompanionService {

    private final TravelCompanionRepository companionRepository;
    private final UserRepository userRepository;

    public List<CompanionDTO> getAllCompanions() {
        return companionRepository.findAll().stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public List<CompanionDTO> searchByDestination(String destination) {
        return companionRepository.findByDestinationNameContainingIgnoreCase(destination).stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public CompanionDTO createCompanion(String username, CompanionDTO dto) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        TravelCompanion companion = TravelCompanion.builder()
                .user(user)
                .destinationName(dto.getDestinationName())
                .travelDates(dto.getTravelDates())
                .interests(dto.getInterests())
                .bio(dto.getBio())
                .contactInfo(dto.getContactInfo())
                .build();

        companion = companionRepository.save(companion);
        return toDTO(companion);
    }

    public void deleteCompanion(Long id) {
        companionRepository.deleteById(id);
    }

    private CompanionDTO toDTO(TravelCompanion c) {
        return CompanionDTO.builder()
                .id(c.getId())
                .userId(c.getUser().getId())
                .username(c.getUser().getUsername())
                .destinationName(c.getDestinationName())
                .travelDates(c.getTravelDates())
                .interests(c.getInterests())
                .bio(c.getBio())
                .contactInfo(c.getContactInfo())
                .build();
    }
}

File: backend/src/services/companion.service.js
const { TravelCompanion } = require('../models');

exports.getAll = async () => {
  const companions = await TravelCompanion.find().populate('userId', 'username');
  return companions.map(c => ({
    id: c._id,
    userId: c.userId?._id,
    username: c.userId?.username,
    destinationName: c.destinationName,
    travelDates: c.travelDates,
    interests: c.interests,
    bio: c.bio,
    contactInfo: c.contactInfo,
    budgetMin: c.budgetMin,
    budgetMax: c.budgetMax
  }));
};

exports.searchByDestination = async (destination) => {
  const regex = new RegExp(destination, 'i');
  const companions = await TravelCompanion.find({ destinationName: regex }).populate('userId', 'username');
  return companions.map(c => ({
    id: c._id,
    userId: c.userId?._id,
    username: c.userId?.username,
    destinationName: c.destinationName,
    travelDates: c.travelDates,
    interests: c.interests,
    bio: c.bio,
    contactInfo: c.contactInfo,
    budgetMin: c.budgetMin,
    budgetMax: c.budgetMax
  }));
};

exports.create = async (userId, data) => {
  const companion = new TravelCompanion({
    userId,
    ...data
  });
  await companion.save();
  return companion;
};

exports.delete = async (id) => {
  const companion = await TravelCompanion.findByIdAndDelete(id);
  if (!companion) {
    throw new Error('Companion not found');
  }
};
